hash/sha1.mbt

```diff
@@ -0,0 +1,285 @@
+// Copyright 2024 International Digital Economy Academy
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file is manually transpiled from the SHA1 C-implementation of
+// the HACL* project:
+// https://github.com/hacl-star/hacl-star/blob/150843809c5fdbc7dac0af395f7a1bb9606f9f96/dist/gcc-compatible/Hacl_Hash_SHA1.c
+// Copyright of the origin:
+// * Copyright (c) 2016-2022 INRIA, CMU and Microsoft Corporation
+// * Copyright (c) 2022-2023 HACL* Contributors
+
+let _h0 = [0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0]
+
+struct SHA1 {
+  block_state : Array[Int]
+  buf : Bytes
+  mut total_len : Int64
+}
+
+pub fn SHA1::make(~init : Bytes = Bytes::make(0, 0)) -> SHA1 {
+  let block_state = _h0.map(fn { x => x })
+  let buf = Bytes::make(64, 0)
+  let total_len = 0L
+  let rv = SHA1::{ block_state, buf, total_len }
+  if init.length() > 0 {
+    rv.update(init)
+  }
+  rv
+}
+
+fn update_multi(
+  h : Array[Int],
+  blocks : @bytes.BytesView,
+  n_blocks : Int
+) -> Unit {
+  for i = 0; i < n_blocks; i = i + 1 {
+    let l = (blocks[i * 64..]) as @mem.Loadable
+    let [ ha, hb, hc, hd, he, .. ] = h
```

Comment:
This is raising a warning that I don't understand:

```
core/hash/sha1.mbt:48:38-48:39 Warning 011: Partial match, some hints:
[ ]
```

Reply:
I assume it means that `h` may be empty